### PR TITLE
Implement identical/equal comparisons on EnumCaseObjectType

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/8485

for comparison see the handling of regular constants/class-constants: https://phpstan.org/r/a6a9d005-a9d5-4f7d-86ad-00a3a3422cec

### CODE REVIEW

File: tests/PHPStan/Rules/Comparison/MatchExpressionRuleTest.php [hunk 1]
```diff
@@ -153,7 +153,11 @@ public function testEnums(): void
 				56,
 			],
 			[
-				'Match arm comparison between *NEVER* and MatchEnums\Foo is always false.',
+				'Match arm comparison between MatchEnums\Foo::THREE and MatchEnums\Foo::THREE is always true.',
```
Comment (staabm) [inline tests/PHPStan/Rules/Comparison/MatchExpressionRuleTest.php:156]: this new error is a bit misleading. not sure we want to keep it like that.

File: tests/PHPStan/Rules/Comparison/MatchExpressionRuleTest.php [hunk 2]
```diff
@@ -153,7 +153,11 @@ public function testEnums(): void
...
+				76,
+			],
+			[
+				'Match arm is unreachable because previous comparison is always true.',
```
Comment (staabm) [inline tests/PHPStan/Rules/Comparison/MatchExpressionRuleTest.php:160]: in contrast, I like this error way more then the one we had before this PR

File: src/Reflection/InitializerExprTypeResolver.php [hunk 3]
```diff
@@ -1303,6 +1303,13 @@ public function resolveIdenticalType(Type $leftType, Type $rightType): BooleanTy
 			return new ConstantBooleanType($leftType->getValue() === $rightType->getValue());
 		}
 
+		if ($leftType instanceof EnumCaseObjectType && $rightType instanceof EnumCaseObjectType) {
+			return new ConstantBooleanType(
```
Comment (ondrejmirtes) [inline src/Reflection/InitializerExprTypeResolver.php]: Use EnumCaseObjectType::equals() instead.

File: tests/PHPStan/Rules/Comparison/MatchExpressionRuleTest.php (hunk 1 above)
Comment (ondrejmirtes) [reply tests/PHPStan/Rules/Comparison/MatchExpressionRuleTest.php:156]: It's fine :)

File: tests/PHPStan/Rules/Comparison/MatchExpressionRuleTest.php (hunk 2 above)
Comment (ondrejmirtes) [reply tests/PHPStan/Rules/Comparison/MatchExpressionRuleTest.php:160]: 👍

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Reflection/InitializerExprTypeResolver.php (hunk 3 above)
Comment (staabm) [reply src/Reflection/InitializerExprTypeResolver.php]: good catch - fixed

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by ondrejmirtes** [comment]
Sorry, I had to revert this, it's pretty annoying in practice. We first need to do a different feature in order for this one to be less annoying.

**Comment by staabm** [comment]
no problem. in case there is something I can do, please open a issue about the thing we need, so I can investigate.

**Comment by ondrejmirtes** [comment]
The problem is that this now reports "always true" for the last `elseif` in an if-elseif-else blocks. In real-world projects this now reports dozens of new errors.

Also arguably last match arm shouldn't report this. Because people might prefer to write `Foo::Bar => doFoo()` instead of `default => doFoo()` to be notified when there's a new enum case to handle.

So, I have an idea: We need a new custom node visitor to mark the condition expression in last match arm and last elseif with an attribute. And ConstantConditionRuleHelper can use this node attribute to decide that "always true" shouldn't be reported there.

There's already an issue for this: https://github.com/phpstan/phpstan/issues/8042

And after this is implemented, we can afford to report all "always true" conditions for everyone in bleedingEdge/2.0, not just phpstan-strict-rules users. Because the most annoying ones will not be reported anymore.

**Pull request merged**